core/kotlinx-coroutines-core/src/test_/TestCoroutineDispatcher.kt

```diff
@@ -0,0 +1,288 @@
+     * under test contains *blocking* Coroutines.
+     *
+     * @param delayTime The amount of time to move the CoroutineContext's clock forward.
+     * @param unit The [TimeUnit] in which [delayTime] and the return value is expressed.
+     * @return The amount of delay-time that this Dispatcher's clock has been forwarded.
+     */
+    fun advanceTimeBy(delayTime: Long, unit: TimeUnit = TimeUnit.MILLISECONDS): Long
+
+    /**
+     * Moves the current virtual clock forward until the next pending delay.
+     *
+     * Any pending immediate dispatches will be executed, and the clock will be advanced to the next delayed dispatch.
+     *
+     * @return the amount of delay-time that this Dispatcher's clock has been forwarded.
+     */
+    fun advanceTimeToNextDelayed(): Long
+
+    /**
+     * Immediately execute all pending tasks and advance the virtual clock-time to the latest delay.
+     *
+     * @return the amount of delay-time that this Dispatcher's clock has been forwarded.
+     */
+    fun runUntilIdle(): Long
+
+    /**
+     * Run any tasks that are pending at or before the current virtual clock-time.
+     *
+     * Calling this function will never advance the clock.
+     */
+    fun runCurrent()
+
+    /**
+     * Call after a test case completes.
+     *
+     * @throws UncompletedCoroutinesError if any pending tasks are active, however it will not throw for suspended
+     * coroutines that called await.
+     */
+    @Throws(UncompletedCoroutinesError::class)
+    fun cleanupTestCoroutines()
+
+    /**
+     * When true, this dispatcher will perform as an immediate executor.
+     *
+     * It will immediately run any tasks, which means it will auto-advance the virtual clock-time to the last pending
+     * delay.
+     *
+     * Test code will rarely call this method directly., Instead use a test builder like [asyncTest], [runBlockingTest] or
+     * the convenience methods [TestCoroutineDispatcher.runBlocking] and [TestCoroutineScope.runBlocking].
+     *
+     * ```
+     * @Test
+     * fun aTest() {
+     *     val scope = TestCoroutineScope() // dispatchImmediately is false
+     *     scope.async {
+     *         // delay will be pending execution (lazy mode)
+     *         delay(1_000)
+     *     }
+     *
+     *     scope.runBlocking {
+     *         // the pending delay will immediately execute
+     *         // dispatchImmediately is true
+     *     }
+     *
+     *     // scope is returned to lazy mode
+     *     // dispatchImmediately is false
+     * }
+     * ```
+     *
+     * Setting this to true will immediately execute any pending tasks and advance the virtual clock-time to the last
+     * pending delay. While true, dispatch will continue to execute immediately, auto-advancing the virtual clock-time.
+     *
+     * Setting it to false will resume lazy execution.
+     */
+    var dispatchImmediately: Boolean
```

Comment:
If you set `dispatchImmediately` to `true` inside an `asyncTest` block, is that effectively like using `runBlockingTest` but with the stricter exception and completion checking?"

Reply:
Yea, I hadn't thought about it that way but it's true. 

Talking out loud: Is there some API simplification that can follow that?

Reply:
Yea that's what I was thinking too - feels like there's some redundancy here.

If you're making `runBlockingTest` do more of the same checks as `asyncTest`, how different are those two builders really, other than just setting this flag differently? Would it make sense to merge them (maybe add a parameter to conveniently set this flag), or at least make it clear in the docs that the only difference is this flag?

Reply:
Ended up resolving this in favor of `runBlockingTest` and more explicit naming here.